### ISSUE TITLE
Allowing arbitrary names for `setinfo` in R

### ISSUE THREAD

**Issue by hetong007** [dmlc/xgboost#1431]
I had a discussion with @road2stat about introducing Cox regression into xgboost. The special thing about it is that in the loss function there are some additional parameters (similar to weights). 

The function `setinfo` enables user to set certain numerical information for each data point in an `xgb.DMatrix` object. In its [current implementation](https://github.com/dmlc/xgboost/blob/master/R-package/R/xgb.DMatrix.R#L240), we only have four options for the parameter `name`. 

@tqchen May I know the reason for that restriction? If we could remove the restriction for the names, Users could attach arbitrary parameters to the data points and it will be more flexible for the loss function implementation.  This helps some applications like Cox regression.

**Comment by road2stat** [comment]
Thanks Tom. It would be great if we could somehow relax this restriction a little. I could also do it by whitelisting new names in `setinfo.xgb.DMatrix` etc., but not sure if it's good practice touching the internal functions.

**Comment by khotilov** [comment]
At R level, you can create a custom objective implemented as a closure that would keep censoring or any other data associated with labels (as @tqchen replied to my suggestion at #574 ).

**Comment by road2stat** [comment]
@khotilov thanks for the hint, do you have any example code, by any chance?

I do think using closure here will help, but I'm not too sure if that's the best way to do this at R level without complicating things unnecessarily.

**Comment by pommedeterresautee** [comment]
Here is an example of custom loss (no use of closure solution, but easy to guess how to do it)
